ADR/ADR037-architecture-of-review-apps.md

```diff
@@ -0,0 +1,120 @@
+        
+sequenceDiagram
+
+actor dev as Dev
+actor reviewer as Reviwer
+participant pr as PR
+participant app as Review App
+
+dev ->> pr: open
+activate pr
+
+pr ->> app: create
+activate app
+
+pr -->> dev: review app URL
+
+loop Iteration
+    dev ->> pr: new commit
+    pr ->> app: redeploy
+    dev ->> app: visit
+end
+
+reviewer ->> pr: review
+pr -->> reviewer: review app URL
+reviewer ->> app: visit
+reviewer ->> pr: approve
+
+dev ->> pr: merge
+pr ->> app: destroy
+deactivate app
+
+pr -->> dev: merged
+deactivate pr
+```
+
+A developer (or non-technical team member) will raise a pull request for their change, and the system we build will deploy a review app in response to the opening of the pull request. The system will share a link to the running app in a comment on the pull request.
+
+Subsequent changes to the pull request will result in the app being redeployed.
+
+Once a reviewer has reviewed and merged the pull request, the system will destroy the running applications. 
+
+### Architecture 
+Given the workflow we want to have, and the investigations we have done, we have decided that we should build a review apps solution ourselves, using common services available to us in AWS. Below is high level architecture diagram for we what we have settled on.
+![A diagram showing the proposed architecture for GOV.UK Forms review apps](../diagrams/images/ADR037-architecture-diagram.png)
+
+We will run the review apps themselves as containers on AWS ECS on Fargate. The task definition we use will deploy containerised PostgreSQL alongside `forms-addmin` and `forms-api`, and configure the apps to talk to the local database. This will ensure that the database in use for review apps is not shared between different instances of the review apps. [^seeding] 
+
+As the interface between GitHub and our own AWS account, we will use [AWS CodeBuild self-hosted GitHub Action runners](https://aws.amazon.com/blogs/devops/aws-codebuild-managed-self-hosted-github-action-runners/). This is a feature of AWS CodeBuild which allows a GitHub Actions workflow to be run on AWS CodeBuild, inside your own AWS account, without needing to provide credentials to GitHub. [^risks] 
+
+The GitHub Actions workflow will build containers and push them to a container registry in the account, and use Terraform to deploy those containers into AWS ECS on Fargate.
+
+Traffic will reach the review apps over two hops:
+1. An ALB will pass the HTTP request to a [Traefik](https://traefik.io/traefik/) proxy.
+2. The Traefik proxy, configured to [discover routing information from AWS ECS](https://doc.traefik.io/traefik/routing/providers/ecs/), will forward the HTTP request onto the relevant container in ECS using host-name based routing.
+
+Using a Traefik proxy will allow us to simplify the Terraform code needed to deploy a review app into AWS by removing the need to add and remove listeners from the ALB.
+
+## Consequences
+
+* We will have a 5th AWS account: integration
```

Comment:
Later on we say we will have a 6th AWS account?

```suggestion
* We will have a 6th AWS account: integration
```

Reply:
These bullet points shouldn't even be here 🤦 They're my leftover notes

Reply:
Oh haha, I thought they were a handy "tell them what you're going to tell them, then tell them" thing haha